// @dimforge/rapier3d/geometry/shape.js
import { VectorOps } from "../math";
import { RawShape } from "../raw";
/**
 * An enumeration representing the type of a shape.
 */
export var ShapeType;
(function (ShapeType) {
    ShapeType[ShapeType["Ball"] = 0] = "Ball";
    ShapeType[ShapeType["Cuboid"] = 1] = "Cuboid";
    ShapeType[ShapeType["Capsule"] = 2] = "Capsule";
    ShapeType[ShapeType["Segment"] = 3] = "Segment";
    ShapeType[ShapeType["Polyline"] = 4] = "Polyline";
    ShapeType[ShapeType["Triangle"] = 5] = "Triangle";
    ShapeType[ShapeType["TriMesh"] = 6] = "TriMesh";
    ShapeType[ShapeType["HeightField"] = 7] = "HeightField";
    // Compound = 8,
    ShapeType[ShapeType["ConvexPolyhedron"] = 9] = "ConvexPolyhedron";
    ShapeType[ShapeType["Cylinder"] = 10] = "Cylinder";
    ShapeType[ShapeType["Cone"] = 11] = "Cone";
    ShapeType[ShapeType["RoundCuboid"] = 12] = "RoundCuboid";
    ShapeType[ShapeType["RoundTriangle"] = 13] = "RoundTriangle";
    ShapeType[ShapeType["RoundCylinder"] = 14] = "RoundCylinder";
    ShapeType[ShapeType["RoundCone"] = 15] = "RoundCone";
    ShapeType[ShapeType["RoundConvexPolyhedron"] = 16] = "RoundConvexPolyhedron";
})(ShapeType || (ShapeType = {}));
// #endif
/**
 * A shape that is a sphere in 3D and a circle in 2D.
 */
var Ball = /** @class */ (function () {
    /**
     * Creates a new ball with the given radius.
     * @param radius - The balls radius.
     */
    function Ball(radius) {
        this.radius = radius;
    }
    Ball.prototype.intoRaw = function () {
        return RawShape.ball(this.radius);
    };
    return Ball;
}());
export { Ball };
/**
 * A shape that is a box in 3D and a rectangle in 2D.
 */
var Cuboid = /** @class */ (function () {
    // #if DIM3
    /**
     * Creates a new 3D cuboid.
     * @param hx - The half width of the cuboid.
     * @param hy - The half height of the cuboid.
     * @param hz - The half depth of the cuboid.
     */
    function Cuboid(hx, hy, hz) {
        this.halfExtents = VectorOps.new(hx, hy, hz);
    }
    // #endif
    Cuboid.prototype.intoRaw = function () {
        // #if DIM3
        return RawShape.cuboid(this.halfExtents.x, this.halfExtents.y, this.halfExtents.z);
        // #endif
    };
    return Cuboid;
}());
export { Cuboid };
/**
 * A shape that is a box in 3D and a rectangle in 2D, with round corners.
 */
var RoundCuboid = /** @class */ (function () {
    // #if DIM3
    /**
     * Creates a new 3D cuboid.
     * @param hx - The half width of the cuboid.
     * @param hy - The half height of the cuboid.
     * @param hz - The half depth of the cuboid.
     * @param borderRadius - The radius of the borders of this cuboid. This will
     *   effectively increase the half-extents of the cuboid by this radius.
     */
    function RoundCuboid(hx, hy, hz, borderRadius) {
        this.halfExtents = VectorOps.new(hx, hy, hz);
        this.borderRadius = borderRadius;
    }
    // #endif
    RoundCuboid.prototype.intoRaw = function () {
        // #if DIM3
        return RawShape.roundCuboid(this.halfExtents.x, this.halfExtents.y, this.halfExtents.z, this.borderRadius);
        // #endif
    };
    return RoundCuboid;
}());
export { RoundCuboid };
/**
 * A shape that is a capsule.
 */
var Capsule = /** @class */ (function () {
    /**
     * Creates a new capsule with the given radius and half-height.
     * @param halfHeight - The balls half-height along the `y` axis.
     * @param radius - The balls radius.
     */
    function Capsule(halfHeight, radius) {
        this.halfHeight = halfHeight;
        this.radius = radius;
    }
    Capsule.prototype.intoRaw = function () {
        return RawShape.capsule(this.halfHeight, this.radius);
    };
    return Capsule;
}());
export { Capsule };
/**
 * A shape that is a segment.
 */
var Segment = /** @class */ (function () {
    /**
     * Creates a new segment shape.
     * @param a - The first point of the segment.
     * @param b - The second point of the segment.
     */
    function Segment(a, b) {
        this.a = a;
        this.b = b;
    }
    Segment.prototype.intoRaw = function () {
        var ra = VectorOps.intoRaw(this.a);
        var rb = VectorOps.intoRaw(this.b);
        var result = RawShape.segment(ra, rb);
        ra.free();
        rb.free();
        return result;
    };
    return Segment;
}());
export { Segment };
/**
 * A shape that is a segment.
 */
var Triangle = /** @class */ (function () {
    /**
     * Creates a new triangle shape.
     *
     * @param a - The first point of the triangle.
     * @param b - The second point of the triangle.
     * @param c - The third point of the triangle.
     */
    function Triangle(a, b, c) {
        this.a = a;
        this.b = b;
        this.c = c;
    }
    Triangle.prototype.intoRaw = function () {
        var ra = VectorOps.intoRaw(this.a);
        var rb = VectorOps.intoRaw(this.b);
        var rc = VectorOps.intoRaw(this.c);
        var result = RawShape.triangle(ra, rb, rc);
        ra.free();
        rb.free();
        rc.free();
        return result;
    };
    return Triangle;
}());
export { Triangle };
/**
 * A shape that is a triangle with round borders and a non-zero thickness.
 */
var RoundTriangle = /** @class */ (function () {
    /**
     * Creates a new triangle shape with round corners.
     *
     * @param a - The first point of the triangle.
     * @param b - The second point of the triangle.
     * @param c - The third point of the triangle.
     * @param borderRadius - The radius of the borders of this triangle. In 3D,
     *   this is also equal to half the thickness of the triangle.
     */
    function RoundTriangle(a, b, c, borderRadius) {
        this.a = a;
        this.b = b;
        this.c = c;
        this.borderRadius = borderRadius;
    }
    RoundTriangle.prototype.intoRaw = function () {
        var ra = VectorOps.intoRaw(this.a);
        var rb = VectorOps.intoRaw(this.b);
        var rc = VectorOps.intoRaw(this.c);
        var result = RawShape.roundTriangle(ra, rb, rc, this.borderRadius);
        ra.free();
        rb.free();
        rc.free();
        return result;
    };
    return RoundTriangle;
}());
export { RoundTriangle };
/**
 * A shape that is a triangle mesh.
 */
var Polyline = /** @class */ (function () {
    /**
     * Creates a new polyline shape.
     *
     * @param vertices - The coordinates of the polyline's vertices.
     * @param indices - The indices of the polyline's segments. If this is `null` then
     *    the vertices are assumed to form a line strip.
     */
    function Polyline(vertices, indices) {
        this.vertices = vertices;
        this.indices = !!indices ? indices : new Uint32Array(0);
    }
    Polyline.prototype.intoRaw = function () {
        return RawShape.polyline(this.vertices, this.indices);
    };
    return Polyline;
}());
export { Polyline };
/**
 * A shape that is a triangle mesh.
 */
var TriMesh = /** @class */ (function () {
    /**
     * Creates a new triangle mesh shape.
     *
     * @param vertices - The coordinates of the triangle mesh's vertices.
     * @param indices - The indices of the triangle mesh's triangles.
     */
    function TriMesh(vertices, indices) {
        this.vertices = vertices;
        this.indices = indices;
    }
    TriMesh.prototype.intoRaw = function () {
        return RawShape.trimesh(this.vertices, this.indices);
    };
    return TriMesh;
}());
export { TriMesh };
// #if DIM3
/**
 * A shape that is a convex polygon.
 */
var ConvexPolyhedron = /** @class */ (function () {
    /**
     * Creates a new convex polygon shape.
     *
     * @param vertices - The coordinates of the convex polygon's vertices.
     * @param indices - The index buffer of this convex mesh. If this is `null`
     *   or `undefined`, the convex-hull of the input vertices will be computed
     *   automatically. Otherwise, it will be assumed that the mesh you provide
     *   is already convex.
     */
    function ConvexPolyhedron(vertices, indices) {
        this.vertices = vertices;
        this.indices = indices;
    }
    ConvexPolyhedron.prototype.intoRaw = function () {
        if (!!this.indices) {
            return RawShape.convexMesh(this.vertices, this.indices);
        }
        else {
            return RawShape.convexHull(this.vertices);
        }
    };
    return ConvexPolyhedron;
}());
export { ConvexPolyhedron };
/**
 * A shape that is a convex polygon.
 */
var RoundConvexPolyhedron = /** @class */ (function () {
    /**
     * Creates a new convex polygon shape.
     *
     * @param vertices - The coordinates of the convex polygon's vertices.
     * @param indices - The index buffer of this convex mesh. If this is `null`
     *   or `undefined`, the convex-hull of the input vertices will be computed
     *   automatically. Otherwise, it will be assumed that the mesh you provide
     *   is already convex.
     * @param borderRadius - The radius of the borders of this convex polyhedron.
     */
    function RoundConvexPolyhedron(vertices, indices, borderRadius) {
        this.vertices = vertices;
        this.indices = indices;
        this.borderRadius = borderRadius;
    }
    RoundConvexPolyhedron.prototype.intoRaw = function () {
        if (!!this.indices) {
            return RawShape.roundConvexMesh(this.vertices, this.indices, this.borderRadius);
        }
        else {
            return RawShape.roundConvexHull(this.vertices, this.borderRadius);
        }
    };
    return RoundConvexPolyhedron;
}());
export { RoundConvexPolyhedron };
/**
 * A shape that is a heightfield.
 */
var Heightfield = /** @class */ (function () {
    /**
     * Creates a new heightfield shape.
     *
     * @param nrows − The number of rows in the heights matrix.
     * @param ncols - The number of columns in the heights matrix.
     * @param heights - The heights of the heightfield along its local `y` axis,
     *                  provided as a matrix stored in column-major order.
     * @param scale - The dimensions of the heightfield's local `x,z` plane.
     */
    function Heightfield(nrows, ncols, heights, scale) {
        this.nrows = nrows;
        this.ncols = ncols;
        this.heights = heights;
        this.scale = scale;
    }
    Heightfield.prototype.intoRaw = function () {
        var rawScale = VectorOps.intoRaw(this.scale);
        var rawShape = RawShape.heightfield(this.nrows, this.ncols, this.heights, rawScale);
        rawScale.free();
        return rawShape;
    };
    return Heightfield;
}());
export { Heightfield };
/**
 * A shape that is a 3D cylinder.
 */
var Cylinder = /** @class */ (function () {
    /**
     * Creates a new cylinder with the given radius and half-height.
     * @param halfHeight - The balls half-height along the `y` axis.
     * @param radius - The balls radius.
     */
    function Cylinder(halfHeight, radius) {
        this.halfHeight = halfHeight;
        this.radius = radius;
    }
    Cylinder.prototype.intoRaw = function () {
        return RawShape.cylinder(this.halfHeight, this.radius);
    };
    return Cylinder;
}());
export { Cylinder };
/**
 * A shape that is a 3D cylinder with round corners.
 */
var RoundCylinder = /** @class */ (function () {
    /**
     * Creates a new cylinder with the given radius and half-height.
     * @param halfHeight - The balls half-height along the `y` axis.
     * @param radius - The balls radius.
     * @param borderRadius - The radius of the borders of this cylinder.
     */
    function RoundCylinder(halfHeight, radius, borderRadius) {
        this.borderRadius = borderRadius;
        this.halfHeight = halfHeight;
        this.radius = radius;
    }
    RoundCylinder.prototype.intoRaw = function () {
        return RawShape.roundCylinder(this.halfHeight, this.radius, this.borderRadius);
    };
    return RoundCylinder;
}());
export { RoundCylinder };
/**
 * A shape that is a 3D cone.
 */
var Cone = /** @class */ (function () {
    /**
     * Creates a new cone with the given radius and half-height.
     * @param halfHeight - The balls half-height along the `y` axis.
     * @param radius - The balls radius.
     */
    function Cone(halfHeight, radius) {
        this.halfHeight = halfHeight;
        this.radius = radius;
    }
    Cone.prototype.intoRaw = function () {
        return RawShape.cone(this.halfHeight, this.radius);
    };
    return Cone;
}());
export { Cone };
/**
 * A shape that is a 3D cone with round corners.
 */
var RoundCone = /** @class */ (function () {
    /**
     * Creates a new cone with the given radius and half-height.
     * @param halfHeight - The balls half-height along the `y` axis.
     * @param radius - The balls radius.
     * @param borderRadius - The radius of the borders of this cone.
     */
    function RoundCone(halfHeight, radius, borderRadius) {
        this.halfHeight = halfHeight;
        this.radius = radius;
        this.borderRadius = borderRadius;
    }
    RoundCone.prototype.intoRaw = function () {
        return RawShape.roundCone(this.halfHeight, this.radius, this.borderRadius);
    };
    return RoundCone;
}());
export { RoundCone };
// #endif
