// @dimforge/rapier3d/dynamics/rigid_body.js
import { RotationOps, VectorOps } from '../math';
/**
 * The simulation status of a rigid-body.
 */
// TODO: rename this to RigidBodyType
export var BodyStatus;
(function (BodyStatus) {
    /**
     * A `BodyStatus::Dynamic` body can be affected by all external forces.
     */
    BodyStatus[BodyStatus["Dynamic"] = 0] = "Dynamic";
    /**
     * A `BodyStatus::Static` body cannot be affected by external forces.
     */
    BodyStatus[BodyStatus["Static"] = 1] = "Static";
    /**
     * A `BodyStatus::KinematicPositionBased` body cannot be affected by any external forces but can be controlled
     * by the user at the position level while keeping realistic one-way interaction with dynamic bodies.
     *
     * One-way interaction means that a kinematic body can push a dynamic body, but a kinematic body
     * cannot be pushed by anything. In other words, the trajectory of a kinematic body can only be
     * modified by the user and is independent from any contact or joint it is involved in.
     */
    BodyStatus[BodyStatus["KinematicPositionBased"] = 2] = "KinematicPositionBased";
    /**
     * A `BodyStatus::KinematicVelocityBased` body cannot be affected by any external forces but can be controlled
     * by the user at the velocity level while keeping realistic one-way interaction with dynamic bodies.
     *
     * One-way interaction means that a kinematic body can push a dynamic body, but a kinematic body
     * cannot be pushed by anything. In other words, the trajectory of a kinematic body can only be
     * modified by the user and is independent from any contact or joint it is involved in.
     */
    BodyStatus[BodyStatus["KinematicVelocityBased"] = 3] = "KinematicVelocityBased";
})(BodyStatus || (BodyStatus = {}));
/**
 * A rigid-body.
 */
var RigidBody = /** @class */ (function () {
    function RigidBody(rawSet, handle) {
        this.rawSet = rawSet;
        this.handle = handle;
    }
    /**
     * Checks if this rigid-body is still valid (i.e. that it has
     * not been deleted from the rigid-body set yet.
     */
    RigidBody.prototype.isValid = function () {
        return this.rawSet.contains(this.handle);
    };
    /**
     * The world-space translation of this rigid-body.
     */
    RigidBody.prototype.translation = function () {
        var res = this.rawSet.rbTranslation(this.handle);
        return VectorOps.fromRaw(res);
    };
    /**
     * The world-space orientation of this rigid-body.
     */
    RigidBody.prototype.rotation = function () {
        var res = this.rawSet.rbRotation(this.handle);
        return RotationOps.fromRaw(res);
    };
    /**
     * The world-space next translation of this rigid-body.
     *
     * If this rigid-body is kinematic this value is set by the `setNextKinematicTranslation`
     * method and is used for estimating the kinematic body velocity at the next timestep.
     * For non-kinematic bodies, this value is currently unspecified.
     */
    RigidBody.prototype.nextTranslation = function () {
        var res = this.rawSet.rbNextTranslation(this.handle);
        return VectorOps.fromRaw(res);
    };
    /**
     * The world-space next orientation of this rigid-body.
     *
     * If this rigid-body is kinematic this value is set by the `setNextKinematicRotation`
     * method and is used for estimating the kinematic body velocity at the next timestep.
     * For non-kinematic bodies, this value is currently unspecified.
     */
    RigidBody.prototype.nextRotation = function () {
        var res = this.rawSet.rbNextRotation(this.handle);
        return RotationOps.fromRaw(res);
    };
    /**
     * Sets the translation of this rigid-body.
     *
     * @param tra - The world-space position of the rigid-body.
     * @param wakeUp - Forces the rigid-body to wake-up so it is properly affected by forces if it
     *                 wasn't moving before modifying its position.
     */
    RigidBody.prototype.setTranslation = function (tra, wakeUp) {
        // #if DIM3
        this.rawSet.rbSetTranslation(this.handle, tra.x, tra.y, tra.z, wakeUp);
        // #endif
    };
    /**
     * Sets the linear velocity fo this rigid-body.
     *
     * @param vel - The linear velocity to set.
     * @param wakeUp - Forces the rigid-body to wake-up if it was asleep.
     */
    RigidBody.prototype.setLinvel = function (vel, wakeUp) {
        var rawVel = VectorOps.intoRaw(vel);
        this.rawSet.rbSetLinvel(this.handle, rawVel, wakeUp);
        rawVel.free();
    };
    /**
     * The scale factor applied to the gravity affecting
     * this rigid-body.
     */
    RigidBody.prototype.gravityScale = function () {
        return this.rawSet.rbGravityScale(this.handle);
    };
    /**
     * Sets the scale factor applied to the gravity affecting
     * this rigid-body.
     *
     * @param factor - The scale factor to set. A value of 0.0 means
     *   that this rigid-body will on longer be affected by gravity.
     * @param wakeUp - Forces the rigid-body to wake-up if it was asleep.
     */
    RigidBody.prototype.setGravityScale = function (factor, wakeUp) {
        this.rawSet.rbSetGravityScale(this.handle, factor, wakeUp);
    };
    // #if DIM3
    /**
     * Sets the rotation quaternion of this rigid-body.
     *
     * This does nothing if a zero quaternion is provided.
     *
     * @param rotation - The rotation to set.
     * @param wakeUp - Forces the rigid-body to wake-up so it is properly affected by forces if it
     * wasn't moving before modifying its position.
     */
    RigidBody.prototype.setRotation = function (rot, wakeUp) {
        this.rawSet.rbSetRotation(this.handle, rot.x, rot.y, rot.z, rot.w, wakeUp);
    };
    /**
     * Sets the angular velocity fo this rigid-body.
     *
     * @param vel - The angular velocity to set.
     * @param wakeUp - Forces the rigid-body to wake-up if it was asleep.
     */
    RigidBody.prototype.setAngvel = function (vel, wakeUp) {
        var rawVel = VectorOps.intoRaw(vel);
        this.rawSet.rbSetAngvel(this.handle, rawVel, wakeUp);
        rawVel.free();
    };
    // #endif
    /**
     * If this rigid body is kinematic, sets its future translation after the next timestep integration.
     *
     * This should be used instead of `rigidBody.setTranslation` to make the dynamic object
     * interacting with this kinematic body behave as expected. Internally, Rapier will compute
     * an artificial velocity for this rigid-body from its current position and its next kinematic
     * position. This velocity will be used to compute forces on dynamic bodies interacting with
     * this body.
     *
     * @param t - The kinematic translation to set.
     */
    RigidBody.prototype.setNextKinematicTranslation = function (t) {
        // #if DIM3
        this.rawSet.rbSetNextKinematicTranslation(this.handle, t.x, t.y, t.z);
        // #endif
    };
    // #if DIM3
    /**
     * If this rigid body is kinematic, sets its future rotation after the next timestep integration.
     *
     * This should be used instead of `rigidBody.setRotation` to make the dynamic object
     * interacting with this kinematic body behave as expected. Internally, Rapier will compute
     * an artificial velocity for this rigid-body from its current position and its next kinematic
     * position. This velocity will be used to compute forces on dynamic bodies interacting with
     * this body.
     *
     * @param rot - The kinematic rotation to set.
     */
    RigidBody.prototype.setNextKinematicRotation = function (rot) {
        this.rawSet.rbSetNextKinematicRotation(this.handle, rot.x, rot.y, rot.z, rot.w);
    };
    // #endif
    /**
     * The linear velocity of this rigid-body.
     */
    RigidBody.prototype.linvel = function () {
        return VectorOps.fromRaw(this.rawSet.rbLinvel(this.handle));
    };
    // #if DIM3
    /**
     * The angular velocity of this rigid-body.
     */
    RigidBody.prototype.angvel = function () {
        return VectorOps.fromRaw(this.rawSet.rbAngvel(this.handle));
    };
    // #endif
    /**
     * The mass of this rigid-body.
     */
    RigidBody.prototype.mass = function () {
        return this.rawSet.rbMass(this.handle);
    };
    /**
     * Put this rigid body to sleep.
     *
     * A sleeping body no longer moves and is no longer simulated by the physics engine unless
     * it is waken up. It can be woken manually with `this.wakeUp()` or automatically due to
     * external forces like contacts.
     */
    RigidBody.prototype.sleep = function () {
        this.rawSet.rbSleep(this.handle);
    };
    /**
     * Wakes this rigid-body up.
     *
     * A dynamic rigid-body that does not move during several consecutive frames will
     * be put to sleep by the physics engine, i.e., it will stop being simulated in order
     * to avoid useless computations.
     * This methods forces a sleeping rigid-body to wake-up. This is useful, e.g., before modifying
     * the position of a dynamic body so that it is properly simulated afterwards.
     */
    RigidBody.prototype.wakeUp = function () {
        this.rawSet.rbWakeUp(this.handle);
    };
    /**
     * Is CCD enabled for this rigid-body?
     */
    RigidBody.prototype.isCcdEnabled = function () {
        this.rawSet.rbIsCcdEnabled(this.handle);
    };
    /**
     * The number of colliders attached to this rigid-body.
     */
    RigidBody.prototype.numColliders = function () {
        return this.rawSet.rbNumColliders(this.handle);
    };
    /**
     * Retrieves the handle of the `i-th` collider attached to this rigid-body.
     *
     * @param i - The index of the collider to retrieve. Must be a number in `[0, this.numColliders()[`.
     *         This index is **not** the same as the unique identifier of the collider.
     */
    RigidBody.prototype.collider = function (i) {
        return this.rawSet.rbCollider(this.handle, i);
    };
    /**
     * The status of this rigid-body: static, dynamic, or kinematic.
     */
    RigidBody.prototype.bodyStatus = function () {
        return this.rawSet.rbBodyStatus(this.handle);
    };
    /**
     * Is this rigid-body sleeping?
     */
    RigidBody.prototype.isSleeping = function () {
        return this.rawSet.rbIsSleeping(this.handle);
    };
    /**
     * Is the velocity of this rigid-body not zero?
     */
    RigidBody.prototype.isMoving = function () {
        return this.rawSet.rbIsMoving(this.handle);
    };
    /**
     * Is this rigid-body static?
     */
    RigidBody.prototype.isStatic = function () {
        return this.rawSet.rbIsStatic(this.handle);
    };
    /**
     * Is this rigid-body kinematic?
     */
    RigidBody.prototype.isKinematic = function () {
        return this.rawSet.rbIsKinematic(this.handle);
    };
    /**
     * Is this rigid-body dynamic?
     */
    RigidBody.prototype.isDynamic = function () {
        return this.rawSet.rbIsDynamic(this.handle);
    };
    /**
     * The linear damping coefficient of this rigid-body.
     */
    RigidBody.prototype.linearDamping = function () {
        return this.rawSet.rbLinearDamping(this.handle);
    };
    /**
     * The angular damping coefficient of this rigid-body.
     */
    RigidBody.prototype.angularDamping = function () {
        return this.rawSet.rbAngularDamping(this.handle);
    };
    /**
     * Sets the linear damping factor applied to this rigid-body.
     *
     * @param factor - The damping factor to set.
     */
    RigidBody.prototype.setLinearDamping = function (factor) {
        this.rawSet.rbSetLinearDamping(this.handle, factor);
    };
    /**
     * Sets the linear damping factor applied to this rigid-body.
     *
     * @param factor - The damping factor to set.
     */
    RigidBody.prototype.setAngularDamping = function (factor) {
        this.rawSet.rbSetAngularDamping(this.handle, factor);
    };
    /**
     * Applies a force at the center-of-mass of this rigid-body.
     *
     * @param force - the world-space force to apply on the rigid-body.
     * @param wakeUp - should the rigid-body be automatically woken-up?
     */
    RigidBody.prototype.applyForce = function (force, wakeUp) {
        var rawForce = VectorOps.intoRaw(force);
        this.rawSet.rbApplyForce(this.handle, rawForce, wakeUp);
        rawForce.free();
    };
    /**
     * Applies an impulse at the center-of-mass of this rigid-body.
     *
     * @param impulse - the world-space impulse to apply on the rigid-body.
     * @param wakeUp - should the rigid-body be automatically woken-up?
     */
    RigidBody.prototype.applyImpulse = function (impulse, wakeUp) {
        var rawImpulse = VectorOps.intoRaw(impulse);
        this.rawSet.rbApplyImpulse(this.handle, rawImpulse, wakeUp);
        rawImpulse.free();
    };
    // #if DIM3
    /**
     * Applies a torque at the center-of-mass of this rigid-body.
     *
     * @param torque - the world-space torque to apply on the rigid-body.
     * @param wakeUp - should the rigid-body be automatically woken-up?
     */
    RigidBody.prototype.applyTorque = function (torque, wakeUp) {
        var rawTorque = VectorOps.intoRaw(torque);
        this.rawSet.rbApplyTorque(this.handle, rawTorque, wakeUp);
        rawTorque.free();
    };
    // #endif
    // #if DIM3
    /**
     * Applies an impulsive torque at the center-of-mass of this rigid-body.
     *
     * @param torqueImpulse - the world-space torque impulse to apply on the rigid-body.
     * @param wakeUp - should the rigid-body be automatically woken-up?
     */
    RigidBody.prototype.applyTorqueImpulse = function (torqueImpulse, wakeUp) {
        var rawTorqueImpulse = VectorOps.intoRaw(torqueImpulse);
        this.rawSet.rbApplyTorqueImpulse(this.handle, rawTorqueImpulse, wakeUp);
        rawTorqueImpulse.free();
    };
    // #endif
    /**
     * Applies a force at the given world-space point of this rigid-body.
     *
     * @param force - the world-space force to apply on the rigid-body.
     * @param point - the world-space point where the impulse is to be applied on the rigid-body.
     * @param wakeUp - should the rigid-body be automatically woken-up?
     */
    RigidBody.prototype.applyForceAtPoint = function (force, point, wakeUp) {
        var rawForce = VectorOps.intoRaw(force);
        var rawPoint = VectorOps.intoRaw(point);
        this.rawSet.rbApplyForceAtPoint(this.handle, rawForce, rawPoint, wakeUp);
        rawForce.free();
        rawPoint.free();
    };
    /**
     * Applies an impulse at the given world-space point of this rigid-body.
     *
     * @param impulse - the world-space impulse to apply on the rigid-body.
     * @param point - the world-space point where the impulse is to be applied on the rigid-body.
     * @param wakeUp - should the rigid-body be automatically woken-up?
     */
    RigidBody.prototype.applyImpulseAtPoint = function (impulse, point, wakeUp) {
        var rawImpulse = VectorOps.intoRaw(impulse);
        var rawPoint = VectorOps.intoRaw(point);
        this.rawSet.rbApplyImpulseAtPoint(this.handle, rawImpulse, rawPoint, wakeUp);
        rawImpulse.free();
        rawPoint.free();
    };
    return RigidBody;
}());
export { RigidBody };
var RigidBodyDesc = /** @class */ (function () {
    function RigidBodyDesc(status) {
        this.status = status;
        this.translation = VectorOps.zeros();
        this.rotation = RotationOps.identity();
        this.gravityScale = 1.0;
        this.linvel = VectorOps.zeros();
        this.mass = 0.0;
        this.translationsEnabled = true;
        this.centerOfMass = VectorOps.zeros();
        // #if DIM3
        this.angvel = VectorOps.zeros();
        this.principalAngularInertia = VectorOps.zeros();
        this.angularInertiaLocalFrame = RotationOps.identity();
        this.rotationsEnabledX = true;
        this.rotationsEnabledY = true;
        this.rotationsEnabledZ = true;
        // #endif
        this.linearDamping = 0.0;
        this.angularDamping = 0.0;
        this.canSleep = true;
        this.ccdEnabled = false;
    }
    /**
     * A rigid-body descriptor used to build a dynamic rigid-body.
     */
    RigidBodyDesc.newDynamic = function () {
        return new RigidBodyDesc(BodyStatus.Dynamic);
    };
    /**
     * A rigid-body descriptor used to build a position-based kinematic rigid-body.
     */
    RigidBodyDesc.newKinematicPositionBased = function () {
        return new RigidBodyDesc(BodyStatus.KinematicPositionBased);
    };
    /**
     * A rigid-body descriptor used to build a velocity-based kinematic rigid-body.
     */
    RigidBodyDesc.newKinematicVelocityBased = function () {
        return new RigidBodyDesc(BodyStatus.KinematicVelocityBased);
    };
    /**
     * A rigid-body descriptor used to build a static rigid-body.
     */
    RigidBodyDesc.newStatic = function () {
        return new RigidBodyDesc(BodyStatus.Static);
    };
    // #if DIM3
    /**
     * Sets the initial translation of the rigid-body to create.
     *
     * @param tra - The translation to set.
     */
    RigidBodyDesc.prototype.setTranslation = function (x, y, z) {
        if (typeof x != "number" || typeof y != "number" || typeof z != "number")
            throw TypeError("The translation components must be numbers.");
        this.translation = { x: x, y: y, z: z };
        return this;
    };
    // #endif
    /**
     * Sets the initial rotation of the rigid-body to create.
     *
     * @param rot - The rotation to set.
     */
    RigidBodyDesc.prototype.setRotation = function (rot) {
        this.rotation = rot;
        return this;
    };
    /**
     * Sets the scale factor applied to the gravity affecting
     * the rigid-body being built.
     *
     * @param scale - The scale factor. Set this to `0.0` if the rigid-body
     *   needs to ignore gravity.
     */
    RigidBodyDesc.prototype.setGravityScale = function (scale) {
        this.gravityScale = scale;
        return this;
    };
    /**
     * Sets the mass of the rigid-body being built.
     *
     * @param mass − The initial mass of the rigid-body to create.
     */
    RigidBodyDesc.prototype.setMass = function (mass) {
        this.mass = mass;
        return this;
    };
    /**
     * Locks all translations that would have resulted from forces on
     * the created rigid-body.
     */
    RigidBodyDesc.prototype.lockTranslations = function () {
        this.translationsEnabled = false;
        return this;
    };
    // #if DIM3
    /**
     * Sets the initial linear velocity of the rigid-body to create.
     *
     * @param x - The linear velocity to set along the `x` axis.
     * @param y - The linear velocity to set along the `y` axis.
     * @param z - The linear velocity to set along the `z` axis.
     */
    RigidBodyDesc.prototype.setLinvel = function (x, y, z) {
        if (typeof x != "number" || typeof y != "number" || typeof z != "number")
            throw TypeError("The linvel components must be numbers.");
        this.linvel = { x: x, y: y, z: z };
        return this;
    };
    /**
     * Sets the initial angular velocity of the rigid-body to create.
     *
     * @param vel - The angular velocity to set.
     */
    RigidBodyDesc.prototype.setAngvel = function (vel) {
        this.angvel = vel;
        return this;
    };
    /**
     * Sets the mass properties of the rigid-body being built.
     *
     * Note that the final mass properties of the rigid-bodies depends
     * on the initial mass-properties of the rigid-body (set by this method)
     * to which is added the contributions of all the colliders with non-zero density
     * attached to this rigid-body.
     *
     * Therefore, if you want your provided mass properties to be the final
     * mass properties of your rigid-body, don't attach colliders to it, or
     * only attach colliders with densities equal to zero.
     *
     * @param mass − The initial mass of the rigid-body to create.
     * @param centerOfMass − The initial center-of-mass of the rigid-body to create.
     * @param principalAngularInertia − The initial principal angular inertia of the rigid-body to create.
     *                                  These are the eigenvalues of the angular inertia matrix.
     * @param angularInertiaLocalFrame − The initial local angular inertia frame of the rigid-body to create.
     *                                   These are the eigenvectors of the angular inertia matrix.
     */
    RigidBodyDesc.prototype.setMassProperties = function (mass, centerOfMass, principalAngularInertia, angularInertiaLocalFrame) {
        this.mass = mass;
        this.centerOfMass = centerOfMass;
        this.principalAngularInertia = principalAngularInertia;
        this.angularInertiaLocalFrame = angularInertiaLocalFrame;
        return this;
    };
    /**
     * Sets the mass properties of the rigid-body being built.
     *
     * @param principalAngularInertia − The initial principal angular inertia of the rigid-body to create.
     */
    RigidBodyDesc.prototype.setPrincipalAngularInertia = function (principalAngularInertia) {
        this.principalAngularInertia = principalAngularInertia;
        return this;
    };
    /**
     * Allow rotation of this rigid-body only along specific axes.
     * @param rotationsEnabledX - Are rotations along the X axis enabled?
     * @param rotationsEnabledY - Are rotations along the y axis enabled?
     * @param rotationsEnabledZ - Are rotations along the Z axis enabled?
     */
    RigidBodyDesc.prototype.restrictRotations = function (rotationsEnabledX, rotationsEnabledY, rotationsEnabledZ) {
        this.rotationsEnabledX = rotationsEnabledX;
        this.rotationsEnabledY = rotationsEnabledY;
        this.rotationsEnabledZ = rotationsEnabledZ;
        return this;
    };
    /**
     * Locks all rotations that would have resulted from forces on
     * the created rigid-body.
     */
    RigidBodyDesc.prototype.lockRotations = function () {
        return this.restrictRotations(false, false, false);
    };
    // #endif
    /**
     * Sets the linear damping of the rigid-body to create.
     *
     * This will progressively slowdown the translational movement of the rigid-body.
     *
     * @param damping - The angular damping coefficient. Should be >= 0. The higher this
     *                  value is, the stronger the translational slowdown will be.
     */
    RigidBodyDesc.prototype.setLinearDamping = function (damping) {
        this.linearDamping = damping;
        return this;
    };
    /**
     * Sets the angular damping of the rigid-body to create.
     *
     * This will progressively slowdown the rotational movement of the rigid-body.
     *
     * @param damping - The angular damping coefficient. Should be >= 0. The higher this
     *                  value is, the stronger the rotational slowdown will be.
     */
    RigidBodyDesc.prototype.setAngularDamping = function (damping) {
        this.angularDamping = damping;
        return this;
    };
    /**
     * Sets whether or not the rigid-body to create can sleep.
     *
     * @param can - true if the rigid-body can sleep, false if it can't.
     */
    RigidBodyDesc.prototype.setCanSleep = function (can) {
        this.canSleep = can;
        return this;
    };
    /**
     * Sets whether Continuous Collision Detection (CCD) is enabled for this rigid-body.
     *
     * @param enabled - true if the rigid-body has CCD enabled.
     */
    RigidBodyDesc.prototype.setCcdEnabled = function (enabled) {
        this.ccdEnabled = enabled;
        return this;
    };
    return RigidBodyDesc;
}());
export { RigidBodyDesc };
